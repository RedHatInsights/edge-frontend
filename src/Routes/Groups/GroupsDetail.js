import React from 'react';
import { useParams } from 'react-router-dom';

const Groups = () => {
    const { uuid } = useParams();
    return <div>Here will be group detail of - {uuid}!</div>;
};

export default Groups;
