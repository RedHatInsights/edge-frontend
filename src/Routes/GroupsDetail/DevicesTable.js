import React, { useState } from 'react';
import { useSelector } from 'react-redux';
import {
  Table,
  TableHeader,
  TableBody,
  sortable,
  cellWidth,
} from '@patternfly/react-table';
import { DateFormat } from '@redhat-cloud-services/frontend-components';
import { StatusIcon } from '../../components';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

const columns = [
  {
    title: 'Name of device',
    transforms: [sortable, cellWidth('max')],
  },
  {
    title: 'Version',
    transforms: [sortable],
  },
  {
    title: 'Last seen',
    transforms: [sortable, cellWidth(15)],
  },
  {
    title: 'Status',
    transforms: [sortable, cellWidth(10)],
  },
];

const DevicesTable = ({ uuid }) => {
  const [sortBy, setSortBy] = useState({});
  const devices = useSelector(
    ({ groupsDetailReducer }) => groupsDetailReducer?.devices
  );
  return (
    <Table
      aria-label="Devices table"
      actions={[
        {
          title: 'Adopt',
          onClick: console.log,
        },
        {
          title: 'Reject',
          onClick: console.log,
        },
      ]}
      cells={columns}
      sortBy={sortBy}
      onSort={(_e, index, direction) =>
        setSortBy(() => ({
          index,
          direction,
        }))
      }
      rows={devices.map((device) => ({
        cells: [
          {
            title: (
              <Link to={`/groups/${uuid}/${device?.uuid}`}>{device?.name}</Link>
            ),
          },
          device?.version,
          {
            title: <DateFormat date={device?.last_seen} />,
          },
          {
            title: <StatusIcon status={device?.status} />,
          },
        ],
      }))}
    >
      <TableHeader />
      <TableBody />
    </Table>
  );
};

DevicesTable.propTypes = {
  uuid: PropTypes.string,
};

export default DevicesTable;
