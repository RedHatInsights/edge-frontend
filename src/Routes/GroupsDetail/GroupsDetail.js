import React, { useState, Suspense } from 'react';
import {
  Breadcrumb,
  BreadcrumbItem,
  Flex,
  FlexItem,
  Skeleton,
  Tooltip,
} from '@patternfly/react-core';
import {
  Dropdown,
  DropdownItem,
  DropdownToggle,
  DropdownPosition,
} from '@patternfly/react-core/deprecated';
import {
  PageHeader,
  PageHeaderTitle,
} from '@redhat-cloud-services/frontend-components/PageHeader';
import Empty from '../../components/Empty';
import { useHistory, useLocation } from 'react-router-dom';
import { routes as paths } from '../../constants/routeMapper';
import CaretDownIcon from '@patternfly/react-icons/dist/esm/icons/caret-down-icon';
import DeviceTable from '../Devices/DeviceTable';
import { useParams } from 'react-router-dom';
import {
  getGroupById,
  removeDeviceFromGroupById,
  removeDevicesFromGroup,
} from '../../api/groups';
import AddSystemsToGroupModal from '../Devices/AddSystemsToGroupModal';
import {
  canUpdateSelectedDevices,
  createLink,
  emptyStateNoFilters,
} from '../../utils';
import useApi from '../../hooks/useApi';
import apiWithToast from '../../utils/apiWithToast';
import { useDispatch } from 'react-redux';
import componentTypes from '@data-driven-forms/react-form-renderer/component-types';
import { Bullseye, Spinner } from '@patternfly/react-core';
import Modal from '../../components/Modal';
import DeleteGroupModal from '../Groups/DeleteGroupModal';
import RenameGroupModal from '../Groups/RenameGroupModal';
import PropTypes from 'prop-types';
import { useFeatureFlags } from '../../utils';
import {
  FEATURE_PARITY_INVENTORY_GROUPS,
  FEATURE_INVENTORY_WORKSPACES_RENAME,
} from '../../constants/features';

const UpdateDeviceModal = React.lazy(() =>
  import(
    /* webpackChunkName: "UpdateDeviceModal" */ '../Devices/UpdateDeviceModal'
  )
);

const GroupsDetail = ({ locationProp, navigateProp, paramsProp }) => {
  const useInventoryGroups = useFeatureFlags(FEATURE_PARITY_INVENTORY_GROUPS);
  const useWorkspacesRename = useFeatureFlags(
    FEATURE_INVENTORY_WORKSPACES_RENAME
  );

  const dispatch = useDispatch();
  const history = useHistory();
  const params = paramsProp ? paramsProp() : useParams();
  const groupId = params?.id ? params.id : params.groupId;
  const { search, pathname } = locationProp ? locationProp() : useLocation();
  const currentApplicationPath =
    window.location.pathname.indexOf('edge') > 0 ? 'edge' : paths.inventory;
  const [isDropdownOpen, setIsDropdownOpen] = useState(false);
  const [isAddModalOpen, setIsAddModalOpen] = useState(false);
  const [removeModal, setRemoveModal] = useState({
    isOpen: false,
    name: '',
    deviceId: null,
  });
  const [updateModal, setUpdateModal] = useState({
    isOpen: false,
    deviceData: null,
    imageData: null,
  });
  const [response, fetchDevices] = useApi({
    api: getGroupById,
    id: groupId,
    tableReload: true,
  });
  const { data, isLoading, hasError } = response;
  const groupName = data?.DeviceGroup?.Name;
  const [deviceIds, getDeviceIds] = useState([]);
  const [hasModalSubmitted, setHasModalSubmitted] = useState(false);
  const [modalState, setModalState] = useState({ id: null, name: '' });
  const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false);
  const [isRenameModalOpen, setIsRenameModalOpen] = useState(false);

  const handleDeleteModal = (id, name) => {
    setModalState({ id, name });
    setIsDeleteModalOpen(true);
  };

  const handleRenameModal = (id, name) => {
    setModalState({ id, name });
    setIsRenameModalOpen(true);
  };

  const removeDeviceLabel = () =>
    `Do you want to remove ${
      deviceIds.length > 0
        ? `${deviceIds.length} system${deviceIds.length === 1 ? '' : 's'}`
        : `${removeModal.name}`
    } from ${groupName}?`;

  const handleSingleDeviceRemoval = () => {
    const statusMessages = {
      onSuccess: {
        title: 'Success',
        description: `${removeModal.name} has been removed successfully`,
      },
      onError: { title: 'Error', description: 'Failed to remove system' },
    };
    apiWithToast(
      dispatch,
      () => removeDeviceFromGroupById(groupId, removeModal.deviceId),
      statusMessages
    );
    setTimeout(() => setHasModalSubmitted(true), 800);
  };

  const handleBulkDeviceRemoval = () => {
    const statusMessages = {
      onSuccess: {
        title: 'Success',
        description: `${deviceIds.length} systems have been removed successfully`,
      },
      onError: { title: 'Error', description: 'failed to remove systems' },
    };
    apiWithToast(
      dispatch,
      () =>
        removeDevicesFromGroup(
          parseInt(groupId),
          deviceIds.map((device) => ({ ID: device.deviceID }))
        ),
      statusMessages
    );
    setTimeout(() => setHasModalSubmitted(true), 800);
  };

  const actionsLabel = 'Actions for group details view';
  const dropdownId = 'group-details-dropdown';

  return (
    <>
      <PageHeader className="pf-m-light">
        {groupName ? (
          <Breadcrumb>
            <BreadcrumbItem>
              {createLink({
                pathname:
                  currentApplicationPath === 'edge'
                    ? `edge${paths.fleetManagement}`
                    : `insights/inventory/groups`,
                linkText: 'Groups',
              })}
            </BreadcrumbItem>
            <BreadcrumbItem>{groupName}</BreadcrumbItem>
          </Breadcrumb>
        ) : (
          <Breadcrumb>
            <BreadcrumbItem isActive>
              <Skeleton width="100px" />
            </BreadcrumbItem>
          </Breadcrumb>
        )}
        <Flex justifyContent={{ default: 'justifyContentSpaceBetween' }}>
          <FlexItem>
            {groupName ? (
              <PageHeaderTitle title={groupName} />
            ) : (
              <Skeleton width="150px" />
            )}
          </FlexItem>
          <FlexItem>
            <Dropdown
              position={DropdownPosition.right}
              toggle={
                <DropdownToggle
                  id={dropdownId}
                  toggleIndicator={CaretDownIcon}
                  onToggle={(_event, newState) => setIsDropdownOpen(newState)}
                  isDisabled={false}
                  aria-label={actionsLabel}
                >
                  <Tooltip
                    content={actionsLabel}
                    triggerRef={() => document.getElementById(dropdownId)}
                  />
                  Actions
                </DropdownToggle>
              }
              isOpen={isDropdownOpen}
              dropdownItems={[
                <DropdownItem
                  key="delete-device-group"
                  onClick={() => handleDeleteModal(groupId, groupName)}
                >
                  Delete group
                </DropdownItem>,
                <DropdownItem
                  key="rename-device-group"
                  onClick={() => handleRenameModal(groupId, groupName)}
                >
                  Rename group
                </DropdownItem>,
                <DropdownItem
                  key="update-all-devices"
                  isDisabled={canUpdateSelectedDevices({
                    deviceData: data?.DevicesView?.devices?.map((device) => ({
                      imageSetId: device?.ImageSetID,
                    })),
                    imageData: data?.DevicesView?.devices?.some(
                      (device) => device.ImageID
                    ),
                  })}
                  onClick={() => {
                    setIsDropdownOpen(false);
                    setUpdateModal((prevState) => ({
                      ...prevState,
                      isOpen: true,
                      deviceData: data?.DevicesView?.devices?.map((device) => ({
                        id: device?.DeviceUUID,
                        display_name:
                          device?.DeviceName === ''
                            ? 'localhost'
                            : device?.DeviceName,
                      })),
                      imageSetId: data?.DevicesView?.devices.find(
                        (device) => device.ImageSetID
                      )?.ImageSetID,
                    }));
                  }}
                >
                  Update
                </DropdownItem>,
              ]}
            />
          </FlexItem>
        </Flex>
      </PageHeader>
      <section className="edge-devices pf-l-page__main-section pf-c-page__main-section">
        {!emptyStateNoFilters(
          isLoading,
          data?.DeviceGroup?.Devices.length,
          search
        ) ? (
          <DeviceTable
            data={data?.DevicesView?.devices || []}
            count={data?.DevicesView?.total}
            isLoading={isLoading}
            locationProp={locationProp}
            navigateProp={navigateProp}
            paramsProp={useParams}
            hasError={hasError}
            hasCheckbox={true}
            handleSingleDeviceRemoval={handleSingleDeviceRemoval}
            kebabItems={[
              {
                isDisabled: !(deviceIds.length > 0),
                title: `Remove from ${
                  useInventoryGroups && useWorkspacesRename
                    ? 'workspace'
                    : 'group'
                }`,
                onClick: () =>
                  setRemoveModal({
                    name: '',
                    deviceId: null,
                    isOpen: true,
                  }),
              },
              {
                isDisabled: canUpdateSelectedDevices({
                  deviceData: deviceIds,
                  imageData: deviceIds[0]?.updateImageData,
                }),
                title: 'Update selected',
                onClick: () =>
                  setUpdateModal((prevState) => ({
                    ...prevState,
                    isOpen: true,
                    deviceData: [...deviceIds],
                    imageSetId: deviceIds.find((device) => device?.imageSetId)
                      .imageSetId,
                  })),
              },
            ]}
            selectedItems={getDeviceIds}
            setRemoveModal={setRemoveModal}
            setIsAddModalOpen={setIsAddModalOpen}
            hasModalSubmitted={hasModalSubmitted}
            setHasModalSubmitted={setHasModalSubmitted}
            fetchDevices={fetchDevices}
            isAddSystemsView={true}
            enforceEdgeGroups={data?.DevicesView?.enforce_edge_groups}
          />
        ) : (
          <Flex justifyContent={{ default: 'justifyContentCenter' }}>
            <Empty
              icon="plus"
              title="Add systems to the group"
              body="Create groups to help manage your systems more effectively."
              primaryAction={{
                text: 'Add systems',
                click: () => setIsAddModalOpen(true),
              }}
              secondaryActions={[
                {
                  type: 'link',
                  title: 'Learn more about system groups',
                  link: 'https://access.redhat.com/documentation/en-us/edge_management/2022/html-single/working_with_systems_in_the_edge_management_application/index',
                },
              ]}
            />
          </Flex>
        )}
      </section>
      {isAddModalOpen && (
        <AddSystemsToGroupModal
          groupId={groupId}
          closeModal={() => setIsAddModalOpen(false)}
          isOpen={isAddModalOpen}
          reloadData={fetchDevices}
          groupName={data?.DeviceGroup?.Name}
          locationProp={locationProp}
          navigateProp={navigateProp}
        />
      )}
      {removeModal.isOpen && (
        <Modal
          isOpen={removeModal.isOpen}
          closeModal={() => setRemoveModal(false)}
          title={`Remove from ${
            useInventoryGroups && useWorkspacesRename ? 'workspace' : 'group'
          }`}
          submitLabel={'Remove'}
          variant="danger"
          schema={{
            fields: [
              {
                component: componentTypes.PLAIN_TEXT,
                name: 'warning-text',
                label: removeDeviceLabel(),
              },
            ],
          }}
          onSubmit={
            removeModal.deviceId
              ? handleSingleDeviceRemoval
              : handleBulkDeviceRemoval
          }
          reloadData={fetchDevices}
        />
      )}

      {updateModal.isOpen && (
        <Suspense
          fallback={
            <Bullseye>
              <Spinner />
            </Bullseye>
          }
        >
          <UpdateDeviceModal
            navigateBack={() => {
              history.push({ pathname });
              setUpdateModal((prevState) => {
                return {
                  ...prevState,
                  isOpen: false,
                };
              });
            }}
            setUpdateModal={setUpdateModal}
            updateModal={updateModal}
            refreshTable={fetchDevices}
          />
        </Suspense>
      )}
      {isDeleteModalOpen && (
        <DeleteGroupModal
          isModalOpen={isDeleteModalOpen}
          setIsModalOpen={setIsDeleteModalOpen}
          reloadData={() => history.push(paths.fleetManagement)}
          modalState={modalState}
        />
      )}
      {isRenameModalOpen && (
        <RenameGroupModal
          isModalOpen={isRenameModalOpen}
          setIsModalOpen={setIsRenameModalOpen}
          reloadData={() => fetchDevices()}
          modalState={modalState}
        />
      )}
    </>
  );
};

GroupsDetail.propTypes = {
  locationProp: PropTypes.func,
  navigateProp: PropTypes.func,
  paramsProp: PropTypes.func,
};

export default GroupsDetail;
