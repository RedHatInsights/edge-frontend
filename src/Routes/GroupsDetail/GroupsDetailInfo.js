import React, { useEffect } from 'react';
import {
  Grid,
  GridItem,
  Card,
  CardBody,
  CardHeader,
  Title,
  Bullseye,
  TextContent,
  Text,
  TextVariants,
  TextListItem,
  Skeleton,
  TextListItemVariants,
  TextList,
  TextListVariants,
} from '@patternfly/react-core';
import { Link } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { loadGroupDevicesInfo } from '../../store/actions';

const GroupsInfo = ({ uuid }) => {
  const dispatch = useDispatch();
  const isDevicesInfoLoading = useSelector(
    ({ groupDevicesInfoReducer }) => groupDevicesInfoReducer?.isLoading
  );
  const devicesInfo = useSelector(
    ({ groupDevicesInfoReducer }) => groupDevicesInfoReducer?.devicesInfo
  );
  useEffect(() => {
    dispatch(loadGroupDevicesInfo());
  }, []);
  return (
    <Grid hasGutter>
      <GridItem span={3}>
        <Card>
          <CardHeader>
            <Title headingLevel="h3">Health threshold</Title>
          </CardHeader>
          <CardBody></CardBody>
        </Card>
      </GridItem>
      <GridItem span={3}>
        <Card>
          <CardHeader>
            <Title headingLevel="h3">Devices info</Title>
          </CardHeader>
          <CardBody>
            <TextContent>
              <TextList component={TextListVariants.dl}>
                <TextListItem component={TextListItemVariants.dt}>
                  {isDevicesInfoLoading === false ? (
                    <Text component={TextVariants.h1}>
                      {devicesInfo?.total}
                    </Text>
                  ) : (
                    <Skeleton />
                  )}
                </TextListItem>
                <TextListItem component={TextListItemVariants.dd}>
                  <Text component={TextVariants.h3}>Total devices</Text>
                </TextListItem>
                <TextListItem component={TextListItemVariants.dt}>
                  {isDevicesInfoLoading === false ? (
                    <Text component={TextVariants.h1}>
                      <Link to={`/groups/${uuid}`}>
                        {devicesInfo?.newDevices}
                      </Link>
                    </Text>
                  ) : (
                    <Skeleton />
                  )}
                </TextListItem>
                <TextListItem component={TextListItemVariants.dd}>
                  New devices added
                </TextListItem>
                <TextListItem component={TextListItemVariants.dt}>
                  {isDevicesInfoLoading === false ? (
                    <Text component={TextVariants.h1}>
                      <Link to={`/groups/${uuid}`}>
                        {devicesInfo?.offlineDevices}
                      </Link>
                    </Text>
                  ) : (
                    <Skeleton />
                  )}
                </TextListItem>
                <TextListItem component={TextListItemVariants.dd}>
                  Devices offline
                </TextListItem>
                <TextListItem component={TextListItemVariants.dt}>
                  {isDevicesInfoLoading === false ? (
                    <Text component={TextVariants.h1}>
                      <Link to={`/groups/${uuid}`}>
                        {devicesInfo?.deliveringDevices}
                      </Link>
                    </Text>
                  ) : (
                    <Skeleton />
                  )}
                </TextListItem>
                <TextListItem component={TextListItemVariants.dd}>
                  Devices on the way
                </TextListItem>
              </TextList>
            </TextContent>
          </CardBody>
        </Card>
      </GridItem>
      <GridItem span={6}>
        <Card className="edge-groups--detail__complex">
          <CardHeader>
            <Grid>
              <GridItem span={6}>
                <Title headingLevel="h3">Canary parameters</Title>
              </GridItem>
              <GridItem span={6}>
                <Title headingLevel="h3">Last Canaries</Title>
              </GridItem>
            </Grid>
          </CardHeader>
          <CardBody>
            <Grid hasGutter>
              <GridItem span={6}>
                <Title headingLevel="h3">
                  <Bullseye>
                    <TextContent>
                      <Text component={TextVariants.h1}>
                        <Bullseye>10% (89)</Bullseye>
                      </Text>
                      <Text component={TextVariants.h4}>
                        <Bullseye>Of systems tested</Bullseye>
                      </Text>
                    </TextContent>
                  </Bullseye>
                </Title>
              </GridItem>
              <GridItem span={6}></GridItem>
            </Grid>
          </CardBody>
        </Card>
      </GridItem>
    </Grid>
  );
};

GroupsInfo.propTypes = {
  uuid: PropTypes.string,
};

export default GroupsInfo;
