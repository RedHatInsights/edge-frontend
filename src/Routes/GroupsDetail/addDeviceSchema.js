import { INVENTORY_TABLE } from '../../components/form';

export default {
  fields: [
    {
      component: INVENTORY_TABLE,
      name: 'select-system',
    },
  ],
};
