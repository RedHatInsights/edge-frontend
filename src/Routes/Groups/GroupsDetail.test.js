import React from 'react';
import GroupsDetail from './GroupsDetail';
import { MemoryRouter, Route } from 'react-router-dom';
import { render } from '@testing-library/react';

describe('Groups', () => {
  it('should render correctly', () => {
    const { container } = render(
      <MemoryRouter initialEntries={['/groups/some-id']}>
        <Route path="/groups/:uuid">
          <GroupsDetail />
        </Route>
      </MemoryRouter>
    );
    expect(container.querySelector('div')).toMatchSnapshot();
  });
});
