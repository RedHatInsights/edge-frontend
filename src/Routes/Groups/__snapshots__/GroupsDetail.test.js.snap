// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Groups should render correctly 1`] = `
<div>
  Here will be group detail of - 
  some-id
  !
</div>
`;
